### PR TITLE
Refactoring & bugfixing for TC #1 & timetable shopping cart

### PR DESCRIPTION
**Pull request by Dante-Capobianco**
In this PR, I implemented the following:

TC#1 bugfixes/refactoring

- Modify normalization function to keep relative positioning & avoiding negative score percentages without having to add all numbers by the "lowest score + 1" if the lowest score is negative (using min-max normalization with an adjustment to make the floor/minimum 30% - configurable parameter)
- Add score additions/deductions (configurable parameters) for each course when it is found in similar user course lists (shopping cart, favorites, removed from favorites, rejected recommendations) such that more than just recommended course scores from similar users is considered (exact matches to these other course lists adds more influence to similar users
- Modify score addition parameter for matches found between courses from similar users & the user of focus' recommended course list (to give more weight to these users)
- Refactor score addition logic for matches between ECE area course lists to avoid repetition across 3 usages of this logic in recommendation function
- **Edge cases/backend test cases handled:**
- Most similar user has a negative score - ignore all similar users due to limited relevance & continue with recommendation logic
- Number of courses outside cart (to be recommended) is under rolling average window length * 2 (the minimum length to apply filtering logic to reduce amount of courses returned to user) - in test case measuring length, ensure recommendation list matches exact length of courses outside of cart, rather than rolling average window length
- No courses are outside cart (no recommendations to give) - return an empty list instead of failing & inform user that no recommendations are available; and in test case looking for a course object list, instead expect any array (to not fail when an empty array is found)
- User clicks through all recommendations on current screen - instead of showing loading more animation infinitely, call recommendation function to retrieve more recommendations (or give message that none are available if all courses added to cart)

Refactoring Logic in Timetable Component

- Refactor shopping cart courses filtering logic to be at top of component, instead of within HTML logic, while still re-rendering on each update of shopping cart
- When user updates courses in shopping cart (adds/removed from cart/favorites across any 3 pages showing it - explore, timetable, shopping cart pages), updates immediately on next render

Frontend 
**1: Last recommendation not dealt with on current returned list of recommendations**
Screenshot 2025-07-14 at 3 41 33 PM

**1: New set of recommendations after deal with last recommendation**
Screenshot 2025-07-14 at 3 41 42 PM

**2: Message when all courses added to cart**
Screenshot 2025-07-14 at 3 41 58 PM

### CODE REVIEW

File: backend/utils/findRecommendedCourses.js [hunk 1]
```diff
@@ -385,6 +394,40 @@ const findMatchesToRelatedUsersCourses = async (
...
+      createIdListFromObjectList(otherUser.favorites)
+    );
+    otherUser.shoppingCart.forEach((cartCourse) => {
+      let courseMatchingOtherUserCart = coursesWithScores.find(
+        (course) => course.id === cartCourse.id
+      );
+      if (courseMatchingOtherUserCart) {
+        courseMatchingOtherUserCart.score +=
+          COURSE_LIKED_ACROSS_SIMILAR_USERS *
+          (favoriteIdList.has(cartCourse.id)
```
Comment (conniewu3) [inline backend/utils/findRecommendedCourses.js:409]: nit: let's extract out logic like this (determining what multiplier to use based on favorites list) into its own variable for easier readability

File: backend/tests/recommendCourses.test.js [hunk 2]
```diff
@@ -1,41 +1,65 @@
...
+  skillsInterests: expect.any(Array),
+  score: expect.any(Number),
 };
 
 // Wording of test title based on documentation (https://jestjs.io/docs/asynchronous)
 test("the recommendation function returns list of course objects with a valid course structure", async () => {
   const courses = await Course.findCourses(userId);
-  const recommendedCourses = await findRecommendedCourses(courses, userId, true);
-  expect(recommendedCourses).toEqual(expect.arrayContaining([expect.objectContaining(expectedCourseFields)]));
+  const cartCourses = await Course.findCoursesInCart(userId);
```
Comment (conniewu3) [inline backend/tests/recommendCourses.test.js:34]: did you have to seed the database with courses for this test to work for userId 1? usually we will populate with fake data to avoid having to pull from the database

File: backend/utils/findRecommendedCourses.js [hunk 3]
```diff
@@ -524,7 +525,7 @@ export const findRecommendedCourses = async (
   });
 
   // If more than two times the rolling average courses are present, keep the top 50% or until a major gap in scores is found
-  if (coursesWithScores.length > NUM_DAYS_ROLLING_AVERAGE * 2) {
+  if (coursesWithScores.length >= NUM_DAYS_ROLLING_AVERAGE * 2) {
```
Comment (conniewu3) [inline backend/utils/findRecommendedCourses.js:528]: nice! did you catch these changes after writing the tests?

File: backend/tests/recommendCourses.test.js (hunk 2 above)
Comment (Dante-Capobianco) [reply backend/tests/recommendCourses.test.js:34]: I did not actually have to do additional seeding (there is simply the original seeding when initializing the db with all of the courses). The findCourses call is done to fetch all courses & the userId is used to simply mark the "inUserShoppingCart" & "inUserFavorites" boolean tag of courses depending on if they are in the user's cart/favorites list or not.

The key purpose of this particular test is a valid course array outcome, which will happen regardless of the user's shopping cart/favorites list. Let me know if this makes sense & if you believe anything should be changed - otherwise, I believe the basic requirement of having a valid user & the initial database seeding is fine.

File: backend/utils/findRecommendedCourses.js (hunk 3 above)
Comment (Dante-Capobianco) [reply backend/utils/findRecommendedCourses.js:528]: I believe so! I noticed this case where at exactly NUM_DAYS_ROLLING_AVERAGE * 2 courses, the logic will still function correctly such that at most half of the courses can successfully move through the filtering logic

File: frontend/src/components/Explore.jsx [hunk 4]
```diff
@@ -46,9 +46,14 @@ const Explore = () => {
...
-      recommendedCourses.filter((course) => course.id !== updatedCourse.id)
+    let newRecommendedCourses = recommendedCourses?.filter(
+      (course) => course.id !== updatedCourse.id
     );
+    if (newRecommendedCourses?.length === 0) {
+      fetchRecommendedCourses();
+    } else {
+      setRecommendedCourses(newRecommendedCourses);
+    }
     fetchAllCourseData();
```
[the remark this answers is not in the document]
Comment (Dante-Capobianco) [reply frontend/src/components/Explore.jsx:57]: This is because this function maps to the Explore Course component, which calls it whenever the user adds to cart, adds to favorite, or rejects a recommendation. Doing so requires not only an update in the recommendations to display (if the user has gone through all current recommendations) but also an update in the explore course list below such that the update (e.g. favorited, added to cart) appears in the course lists immediately.
